### ISSUE TITLE
Does no registration count as a unique registration? 

### ISSUE THREAD

**Issue by garemoko** [adlnet/xapi-profiles#40]
> A pattern is determined by a given registration — all statements within a Pattern MUST use the same registration, and statements not part of a Pattern MUST NOT use the same registration as any that are.

Could a LRP satisfy this requirement by not using the registration property at all but meeting the requirements of the pattern across all statements for each learner? (i.e. the uniqueness of the actor replaces the uniqueness of the registration).

**Comment by fugu13** [comment]
I don't think no registration should count. The model of "here's a discrete window of events you need to describe" doesn't really expand to "here's all this person's events related to this profile ever" -- imagine someone does a cmi5 course ten times, but one of the times the course was written badly and there's a mess-up. Even if we make rules for starting new patterns without new registrations somehow, how do we toss the data for just the one messed up taking and not the other nine?

I'd also like to not do joint actor+registration unique (though that's somewhat a practice right now). If we clearly mark sessions as being unique registrations, we gain the capability to have sessions that involve multiple people.

**Comment by garemoko** [comment]
The use case I have in mind here is using the profile spec to describe statements sent by a particular product e.g. an authoring tool or a video platform. Since registration is an optional property of xAPI, requiring registration for all patterns prevents patterns being used for that use case. I do think patterns could be valuable in those cases even when registration is not used. 

Yes, the lack of registration would make validation of patterns difficult, however tools could still determine the start and end of a pattern instance by matching the relevant statement templates to the first and last template.

We could still include a requireRegistration property for patterns and recommend it's true for most profiles, even saying that the profile server doesn't support validation of profiles without it if that helps.

**Comment by fugu13** [comment]
I think you're talking about using profiles to describe the behavior of a tool that has already been written or is written without concern for the profile spec? If so, I'd be pretty against trying to provide extensive support for that; such a tool could literally do anything xAPI legal, and we'd quickly go down a hole of "okay, so we'll add this new flag to accommodate weird tool behavior Z as an option in profiles".

That doesn't mean that people couldn't experiment with ways to "lift" non-profile data into profile-workable data, but I think trying to make the profile spec do that is a mistake.

I'd like to particularly point out that the existence of clear first and last templates is very far from a given, re:

> Yes, the lack of registration would make validation of patterns difficult, however tools could still determine the start and end of a pattern instance by matching the relevant statement templates to the first and last template.

Plus, that's adding yet another piece of otherwise extraneous logic to accommodate non-profile-smart tools.

Worth discussing, but we should be aware this is a *huge* rabbit hole.

**Comment by garemoko** [comment]
> I think you're talking about using profiles to describe the behavior of a tool that has already been written or is written without concern for the profile spec.

Yes. This is the main way I intend to use profiles and to me being able to document existing practice in a sharable way is considerably more valuable than ensuring that all profiles can be easily validated. I hope that we can find a middle ground that allows for both though. Maybe we state that profile servers are not required to be able to validate patterns in statements that both a) do not include registration and b) follow a profile that does not require registration for the pattern. 

I'm not saying that profiles themselves shouldn't have restrictions in there that are incompatible with existing practice, of course they will, I just think that the profile spec should be able describing xAPI statements rather than restricting them; leave the restrictions to the profiles themselves.

**Comment by fugu13** [comment]
You can absolutely look through registrationless statements and see if they otherwise match a profile -- there's nothing in the profile spec stopping you right now. All the profile spec says is, if a tool wants to send statements according to the rules for profiles and expect profile-aware systems to understand that, it must include the profile in the categories and include registration for statements sent under one pattern.

But I don't see any need for the profile spec to include options to say that "statements might not be sent in a way that makes these profiles usable for understanding them" -- the right ways to use the profile spec should all have solid use for processing data, not just describing it.

And it would make it virtually useless for processing data. Imagine a tool sends statements according to this pattern:

`(B | C) A* (C A*)?`

With registrations, figuring out if the pattern is matched is easy. Without registrations, it is completely undecidable where a new pattern starts vs where the optional part of the pattern starts, if someone does the pattern multiple times.

And the thing is, we can't say "just don't write those patterns" -- first, turns out figuring out if it is one of those patterns is a bit tricky, but more importantly, the scenario you're talking about is descriptive -- whatever the pattern of statements the tool sends is the pattern of statements the tool sends, and needs to be described, even if has these problems.

But okay, lets imagine all patterns are simpler. Here's a really simple one: `B A B`, no optionality, just a short sequence of statements that will always be that way -- perhaps describing a series of statements seen for an interaction of mild complexity. With or without registrations, seems very simple... until we have a data error. We're looking at a series of statements and we see: `B A B B A B B A B B A B B A B B A B B A A B A B B A B B A B B A B`

See in the middle where it goes `B A A B A B`? Without registrations there's no way to tell which parts are corrupt and which parts are real... and the real answer might well be, if there were substantial network problems, that actually the problems started three statements before that, which are all from separate interactions that just happened to arrive in a way that looks like a good pattern. With registrations, the determination is trivial.

In other words, if we include a "no registration" flag, what we're including is "if you set this flag, no system can reliably be expected to make sense of your statements in a way consistent with how other systems make sense of them, so this part of the profile's just there for humans to read and not there for machines to work with" -- and you can already just write a textual description of those statements for humans.

(It would still be useful for authoring, except the scenario given is a tool that's already been authored -- and anyone using the profile for authoring would be authoring data that would have all the problems just given, which is a very bad idea).

**Comment by garemoko** [comment]
I agree that delimiting patterns of statements without a registration is difficult and problematic. I disagree that this means there is no value to attempting to do so and getting it right some of the time. 

We have exactly this problem in Watershed were some report configurations work around looking at patterns of statements. For exactly the reasons you've described, this is much more accurate where a registration property is provided and we encourage our customers and partners to include a registration where possible. 

But the world is messy and sometimes we do have to deal with statements that don't include registration. In those cases Watershed makes best guesses and assumptions to calculate some interpretation. It might not be exactly right but it still has value as an approximation. 

> "if you set this flag, no system can reliably be expected to make sense of your statements in a way consistent with how other systems make sense of them, so this part of the profile's just there for humans to read and not there for machines to work with"

I again agree with the first part - no system can reliably make sense of your statements - but disagree that this means it's not there for machines. In fact, I want to allow for profiles to allow no-registration statements precisely because of the value to Watershed, an application. Machines can make guesses and approximations too. See [this great example of that I came across this morning](https://www.autodraw.com/).

I'm not saying that we encourage statements with no registration, we can strongly recommend that LRPs use context registration, and we can strongly recommend that profile authors absolutely require registration for conformance to their profiles. I'm just suggesting we leave the door open for profile authors to make the decision for themselves.

**Comment by fugu13** [comment]
Interesting scenario brought up in discussion:

Person takes course. Course contains two videos. Current practice and one we want to encourage is put the same registration on all statements. How do we distinguish the two video segments -- their patterns -- when they have the same registration? Do we require different registrations? Lots of open questions.

**Comment by fugu13** [comment]
Brought up in discussion:

Possible future change for xAPI Version Breaking.0: array for registration.

Possible adaptation by profile spec statements: a context extension allowing registration-like arrays.

**Comment by fugu13** [comment]
Due to complicated internal issue identified above, we'll circle back to the larger issue at a later meeting.

**Comment by fugu13** [comment]
It sounds like people like "self-punctuating" patterns, where the 'start' and 'end' statement templates available in a pattern are exclusive with each other and with the statement templates in the 'middle'.

**Comment by garemoko** [comment]
One issue with "self-punctuating" patterns raised on today's call was that the learner might have two sessions running at the same time. 

This led to use discussing the 'come up with new extensions' option. We discussed the possibility of using registration to differentiate in simple cases where there are no instances of things within things (videos in e-learning courses), with the extension used to document child and parent registrations of that main registration that could be used to differentiate between patterns involving things that are within other things.

**Comment by fugu13** [comment]
That's definitely an issue. Here's a sketch of a possible approach:

1. Registration must still be used.
2. When only one profile or only one pattern occurrence of each profile will be used, only registration SHOULD be used.
3. The extension to be used (MUST for multiple pattern occurrences, MAY for single profile occurrences) to clarify "sub registrations" (probably not the term we want to use but works for now) looks like:

```javascript
[
  {
    "profile": "http://example.org/profile/1",
    "subregistration": "insert uuid here"
   },
   {
    "profile": "http://example.org/profile/2",
    "subregistration": "another uuid here"
   }
]
```

Within a given registration + sub registration combination for a given profile, there would be expected to be exactly one complete pattern.

There can be multiple because a single statement can adhere to multiple profiles.

Thoughts?

**Comment by garemoko** [comment]
Sounds about right (although I'm still in favor of allowing profiles not to require registration). 

Do we also need to include a pattern id to cover cases where a profile include multiple patterns that could overlap?

**Comment by garemoko** [comment]
Here's another interesting edge case. gomo authoring tool (and maybe others) allows you to embed an e-learning course in to a slide. You could therefore potentially have two (or more) versions of the same pattern running at the same time, with the same registration. 

I *think* that still works with the subregiustration approach, which makes me realize that we don't need the "profile" property. The extension could just be an array of UUIDs. I think?

**Comment by fugu13** [comment]
I think the gomo scenario works okay -- each statement will still only be for one subregistration of a given profile, even if they're contained in the same top-level registration.

**Comment by fugu13** [comment]
Also, the same structure would allow for multiple patterns to be progressing for the same profile, overlapping, come to think of it. Though I think a lot of profiles would probably define statement templates such that that's less easy, for ambiguity reasons. For example, I don't think it would be feasible to actually overlap subregistrations  of the scorm profile example as I gave, only to nest separate ones within the same profile.

**Comment by aaronesilvers** [comment]
As this issue was opened two months ago, I would ask the community to consider the approach to require registration, as @fugu13 mentioned above. For our 12 June meeting, if there are use cases other than finding patterns in legacy statements without registrations to *not* do this, please share them next week. Otherwise, we intend to require this and contribute a PR to close this issue.

```
[
  {
    "profile": "http://example.org/profile/1",
    "subregistration": "insert uuid here"
   },
   {
    "profile": "http://example.org/profile/2",
    "subregistration": "another uuid here"
   }
]
```

**Comment by fugu13** [comment]
Current idea: add a note to the profile spec that some profiles might describe patterns that old data adhered to, but not as strictly, and that handling the old data is outside the scope of the profile spec, but might be useful in some contexts. Recommend that profiles with nearly-matching old data likely out there include a note to the effect, including what ways the old data is likely not to comply. Warn people that handling the old data will always be somewhat heuristic, so any new profile adopters MUST follow the rules strictly.
